### PR TITLE
nixos/matter-server: init

### PR DESCRIPTION
**Pull request by leonm1**
## Description of changes

This PR adds a new module for executing `python-matter-server` as a systemd service.

## Things done

1. Patched python-matter-server not not require internet access to download certificates
1. Added nixos module to run python-matter-server as a systemd service
1. Added tests that the matter-server service can start up and respond to websocket connections

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by leonm1** [comment]
@mweinelt

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:14]: That should be you, not me.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:24]:
```suggestion
    package = mkPackageOptionMD pkgs "python-matter-server" { };
```

Shorthand for the package option.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:29]:
```suggestion
```

Only required when the default value is a complex expression or references other values.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:36]:
```suggestion
```

Same as above. Is there any point in offering the user a choice for `vendorId` and `fabricId`? We don't need to expose options prominently, if nobody is going to make use of them

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:43]: Mismatch, but `defaultText` is redundant anyway.

```suggestion
      default = 1;
```

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:51]:
```suggestion
```

Instead of exposing a storage-path, let's drop this option and people who want to store their state elsewhere can configure a bind mount on top of `/var/lib/matter-server`.

Systemd handles mounts on top of those directories extremely well, e.g. by setting up a dependency for the mount point on the unit.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:72]:
```suggestion
        "avahi-daemon.service"
```

I'm surprised it wants an avahi-daemon, can you reference the documentation where it says so?

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:73]:
```suggestion
      after = [
        "network-online.target"
        "avahi-daemon.target"
      ];
      wants = [
        "network-online.target"
        "avahi-daemon.target"
      ];
```

Will be required once `multi-user.target` does not want `network-online.target` anymore. so let's set this up correctly now.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:76]:
```suggestion
        "home-assistant.target"
```

Preferably via `wants` and `before`, because if `home-assistant.target` isn't reached, you'll still want `mattter-server.service` to boot up.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:90]: Can you explain the idea behind this?

File: nixos/modules/services/home-automation/matter-server.nix
Comment (mweinelt) [inline nixos/modules/services/home-automation/matter-server.nix:114]:
```suggestion
        RestrictAddressFamilies = [
        	"AF_INET"
        	"AF_INET6"
        	"AF_UNIX"
        	"AF_NETLINK"
        	"AF_BLUETOOTH"
        ];
```

Expanding lists makes them better readable and more robust when a merge is required.

I'm somewhat surprised by `AF_UNIX`, `AF_NETLINK` and `AF_BLUETOOTH`, since my idea of the matter-server was, that it supported either Thread or WLAN.

**Review by mweinelt — DISMISSED** [review]
I did a first pass and things are looking pretty good already.

Can you get me some logs how python-matter-server 4.0 failed for you? This is something we need to investigate before home-assistant 2023.11.0 starts depending on it on November 1st.

And if you do a revert, the commit message should contain the reasoning behind the revert.

File: nixos/tests/matter-server.nix
Comment (h7x4) [inline nixos/tests/matter-server.nix:37]: nested subtests for only a single assert seems overkill. Let's just add the message to the assert itself?

```suggestion
        assert '"sdk_version": "${chipVersion}"' in output, (
          "CHIP version \"${chipVersion}\" not present in websocket message"
        )

        assert '"fabric_id": 7654' in output, (
            "fabric_id not propagated to server"
        )
```

File: nixos/tests/matter-server.nix
Comment (mweinelt) [reply nixos/tests/matter-server.nix:22]: I think this is a marker for an editor to highlight this as python. I might be wrong.

File: maintainers/maintainer-list.nix
Comment (Janik-Haag) [inline maintainers/maintainer-list.nix:11532]: please split this out into it's own commit called: `maintainers: add mattleon`

Comment (dotlambda) [reply maintainers/maintainer-list.nix:11532]: And I think
```suggestion
  leonm1 = {
    github = "leonm1";
    githubId = 32306579;
    keys = [{
      fingerprint = "C12D F14B DC9D 64E1 44C3  4D8A 755C DA4E 5923 416A";
    }];
    matrix = "@mattleon:matrix.org";
    name = "Matt Leon";
  };
```
is preferred.

Comment (mweinelt) [reply maintainers/maintainer-list.nix:11532]: We are not telling people what maintainer name they have to choose, so please don't.

Comment (dotlambda) [reply maintainers/maintainer-list.nix:11532]: https://github.com/NixOS/nixpkgs/blob/684e1e86e2eff075300c5e59b4588e80e7ab188f/maintainers/maintainer-list.nix#L34

Comment (Janik-Haag) [reply maintainers/maintainer-list.nix:11532]: But if they already decided against that we don't have to ask them again.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Done.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Thanks for the suggestion! TIL.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Thanks for the suggestion! TIL.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Done.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Done.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Good simplification, thanks for the suggestion.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Done. TIL.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: I have added comments to explain the filesystem setup. PTAL and let me know if there's anything I can clarify.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: This appeared to be my mistake — it works just fine without avahi-daemon, which makes sense, since it can broadcast via DNS-SD without avahi. Removed.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Done.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Good call-out. AF_UNIX sockets were not required, nor does it seem AF_BLUETOOTH.

Blocking AF_NETLINK sockets results in this test failure:

```
machine # [   13.109288] matter-server[951]: 2024-01-15 20:56:52 machine chip.DL[951] ERROR Failed to init netlink socket for IP addresses: 97
machine # [   13.111273] matter-server[951]: Traceback (most recent call last):
machine # [   13.112619] matter-server[951]:   File "/nix/store/h985xfmvi4dfpq51z2pa95d0k3mvvg6n-python3.11-python-matter-server-5.1.1/bin/.matter-server-wrapped", line 9, in <module>
machine # [   13.115242] matter-server[951]:     sys.exit(main())
machine # [   13.116517] matter-server[951]:              ^^^^^^
machine # [   13.117864] matter-server[951]:   File "/nix/store/h985xfmvi4dfpq51z2pa95d0k3mvvg6n-python3.11-python-matter-server-5.1.1/lib/python3.11/site-packages/matter_server/server/__main__.py", line 82, in main
machine # [   13.120922] matter-server[951]:     server = MatterServer(
machine # [   13.128743] matter-server[951]:              ^^^^^^^^^^^^^
machine # [   13.131310] matter-server[951]:   File "/nix/store/h985xfmvi4dfpq51z2pa95d0k3mvvg6n-python3.11-python-matter-server-5.1.1/lib/python3.11/site-packages/matter_server/server/server.py", line 74, in __init__
machine # [   13.138086] matter-server[951]:     self.stack = MatterStack(self)
machine # [   13.140959] matter-server[951]:                  ^^^^^^^^^^^^^^^^^
machine # [   13.144398] matter-server[951]:   File "/nix/store/h985xfmvi4dfpq51z2pa95d0k3mvvg6n-python3.11-python-matter-server-5.1.1/lib/python3.11/site-packages/matter_server/server/stack.py", line 32, in __init__
machine # [   13.152656] matter-server[951]:     self._chip_stack = ChipStack(
machine # [   13.156391] matter-server[951]:                        ^^^^^^^^^^
machine # [   13.159454] matter-server[951]:   File "/nix/store/sp558r7pil10npl538dan75gqkpw56f9-python3.11-home-assistant-chip-core-2023.12.0/lib/python3.11/site-packages/chip/ChipStack.py", line 64, in wrapper
machine # [   13.166879] matter-server[951]:     instance[0] = cls(*args, **kwargs)
machine # [   13.171054] matter-server[951]:                   ^^^^^^^^^^^^^^^^^^^^
machine # [   13.174143] matter-server[951]:   File "/nix/store/sp558r7pil10npl538dan75gqkpw56f9-python3.11-home-assistant-chip-core-2023.12.0/lib/python3.11/site-packages/chip/ChipStack.py", line 271, in __init__
machine # [   13.182048] matter-server[951]:     res.raise_on_error()
machine # [   13.184846] matter-server[951]:   File "/nix/store/sp558r7pil10npl538dan75gqkpw56f9-python3.11-home-assistant-chip-core-2023.12.0/lib/python3.11/site-packages/chip/native/__init__.py", line 67, in raise_on_error
machine # [   13.192148] matter-server[951]:     raise self.to_exception()
machine # [   13.199982] matter-server[951]: chip.exceptions.ChipStackError: src/platform/Linux/PlatformManagerImpl.cpp:165: Error 0x000000AC


```

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Done.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Done.

File: pkgs/development/python-modules/python-matter-server/default.nix
Comment (leonm1) [reply pkgs/development/python-modules/python-matter-server/default.nix]: Done.

File: nixos/tests/matter-server.nix
Comment (leonm1) [reply nixos/tests/matter-server.nix]: Done. Thanks for the suggestion — I successfully evade python most of the time, so I was not aware of this syntax.

File: maintainers/maintainer-list.nix
Comment (leonm1) [reply maintainers/maintainer-list.nix]: Made a separate commit to follow [pkgs/README.md#commit-conventions](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions). I updated my maintainer handle to follow the recommendation, since I understand that can be confusing to some and don't want to cause confusion.

File: nixos/tests/matter-server.nix
Comment (leonm1) [reply nixos/tests/matter-server.nix:20]: This is indeed an editor guide to syntax highlight the nix string. Please let me know if I should remove it still.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: This is now obsolete.

**Comment by meutraa** [comment]
I'm not having much luck testing this. Devices report in android that something went wrong when commissioning, but matter-server logs show them succeeding. Then I get the device resubscribing every few seconds. Adding a second device something gets confused and home-assistant only reports one device that toggles both devices.

**Comment by leonm1** [comment]
> I'm not having much luck testing this. Devices report in android that something went wrong when commissioning, but matter-server logs show them succeeding. Then I get the device resubscribing every few seconds. Adding a second device something gets confused and home-assistant only reports one device that toggles both devices.

I troubleshooted this tonight, and was able to reproduce it with my nanoleaf holiday string lights. Digging into the debug logs (`config.services.matter-server.logLevel = "debug";`), I received an error in the commissioning step `SendNOC`, with the error logged [here](https://github.com/project-chip/connectedhomeip/blob/181b0cb14ff007ec912f2ba6627e05dfb066c008/src/controller/CHIPDeviceController.cpp#L1441C16-L1441C40) in the chip project. 

The error was that the device was refusing to join a "fabic" that it already considered itself connected to. AFAICT, fabrics are the matter protocol way for the device to join multiple different home automation platforms, so when the device thought it was already attached to home-assistant's "fabric", it refused to join again. Resetting my device made the error go away.

Are you by any chance using nanoleaf devices?

**Comment by meutraa** [comment]
> > I'm not having much luck testing this. Devices report in android that something went wrong when commissioning, but matter-server logs show them succeeding. Then I get the device resubscribing every few seconds. Adding a second device something gets confused and home-assistant only reports one device that toggles both devices.
> 
> I troubleshooted this tonight, and was able to reproduce it with my nanoleaf holiday string lights. Digging into the debug logs (`config.services.matter-server.logLevel = "debug";`), I received an error in the commissioning step `SendNOC`, with the error logged [here](https://github.com/project-chip/connectedhomeip/blob/181b0cb14ff007ec912f2ba6627e05dfb066c008/src/controller/CHIPDeviceController.cpp#L1441C16-L1441C40) in the chip project.
> 
> The error was that the device was refusing to join a "fabic" that it already considered itself connected to. AFAICT, fabrics are the matter protocol way for the device to join multiple different home automation platforms, so when the device thought it was already attached to home-assistant's "fabric", it refused to join again. After factory resetting my device by [holding the power button for 10s](https://helpdesk.nanoleaf.me/en-US/how-to-control-your-holiday-string-lights-using-button-342048), I was able to commission it flawlessly. (Note that deleting in the nanoleaf / home-assistant apps did not work, but factory reset did).
> 
> Are you by any chance using nanoleaf devices?

No, I am using Tapo 110M, Some Phillips Wiz light bulbs, and an Athom relay switch. I'm testing the commissions currently with the Tapo devices since they are easy to factory reset each time. 

The debug log is fairly massive, with this one, I factory reset a tapo 110m, commissioned it (something went wrong), though matter-server's log seems to report it working. Then repeated with a second 110m, this time matter-server reporting an error.

[main.log](https://github.com/NixOS/nixpkgs/files/13948816/main.log)

Although, this isn't always the case, sometimes they add, but home-assistant reports one device yet controls many seemingly at the same time.

I'm not sure if my network has anything to do with it, since a previous instance of home-assistant os running on a pi4 commissioned many no problem, as does google home. But just in case, my network configuration for nixos is
[network.txt](https://github.com/NixOS/nixpkgs/files/13948888/network.txt) and the resulting network 
[ip.txt](https://github.com/NixOS/nixpkgs/files/13948914/ip.txt)

I tried without avahi enabled but I'm unable to commission any devices with a timeout on mdns.

**Comment by meutraa** [comment]
Looking at this issue: https://github.com/home-assistant-libs/python-matter-server/issues/463#issuecomment-1877034595 my issue might not be related to the packaging.

File: nixos/tests/matter-server.nix
Comment (h7x4) [reply nixos/tests/matter-server.nix:20]: Feel free to leave it if you'd like. I just didn't see the value of it initially.

**Comment by leonm1** [comment]
@mweinelt could you take another look at this PR? I think it's ready to merge.

File: pkgs/development/python-modules/python-matter-server/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/python-matter-server/default.nix:33]:
```suggestion
    pname = "matter-server-paa-certificates";
    version = "1.2.0.1";
```

File: pkgs/development/python-modules/python-matter-server/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/python-matter-server/default.nix:38]:
```suggestion
      rev = "refs/tags/v${version}";
```

File: nixos/modules/services/home-automation/matter-server.nix
Comment (SuperSandro2000) [inline nixos/modules/services/home-automation/matter-server.nix:56]:
```suggestion
      after = [ "network-online.target" ];
      before = [ "home-assistant.service" ];
      wants = [ "network-online.target" ];
      wantedBy = [ "multi-user.target" ];
```
nit but saves some space

File: nixos/modules/services/home-automation/matter-server.nix
Comment (SuperSandro2000) [inline nixos/modules/services/home-automation/matter-server.nix:39]: Would be cool if we could link some upstream doc

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2405.section.md:69]:
```suggestion
- [Python Matter Server](https://github.com/home-assistant-libs/python-matter-server), a
  Matter Controller Server exposing websocket connections for use with other services, notably Home Assistant.
  Available as [services.matter-server](#opt-services.matter-server.enable)
```
keeping links in one line makes them easy searchable. Also we don#t need to adhear to such short lines. They usually just make things complicated. Doing one sentence per line is a pretty good alternative.

File: pkgs/development/python-modules/python-matter-server/default.nix
Comment (leonm1) [reply pkgs/development/python-modules/python-matter-server/default.nix]: Thanks for pointing this out, I had to search what the difference between pname and name is, but I think I understand now.

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Done. I left this in one line, so it's pretty long now. Is that how it is preferred?

File: nixos/modules/services/home-automation/matter-server.nix
Comment (SuperSandro2000) [inline nixos/modules/services/home-automation/matter-server.nix:39]:
```suggestion
      description = ''
        Extra arguments to pass to the matter-server executable.
        See https://github.com/home-assistant-libs/python-matter-server?tab=readme-ov-file#running-the-development-server for options.
      '';
```
as you have noticed in the other thread already that this line is not prety long but I think it is quite easy to fix that here.

**Review by SuperSandro2000 — APPROVED** [review]
LGTM

File: nixos/modules/services/home-automation/matter-server.nix
Comment (leonm1) [reply nixos/modules/services/home-automation/matter-server.nix]: Done. Thanks for your review!

**Comment by K900** [comment]
Literally does not build.

**Comment by SuperSandro2000** [comment]
ofborg can't build it because it uses opens so 1.1.1

**Comment by mweinelt** [comment]
Correct, but IMO that makes an even stronger case that you should actually build/test things before merging them.

**Pull request merged**
